File: TimeFlies/TimeFliesView.m
//
//  TimeFliesView.m
//  TimeFlies
//
//  Created by Vergil Choi on 2018/4/9.
//  Copyright © 2018 Vergil Choi. All rights reserved.
//

#import "TimeFliesView.h"

@import AppKit;

@interface TimeFliesView()

@property (nonatomic, strong) NSTextField *customLabel;

@property (nonatomic, strong) NSTextField *percentNumber;
@property (nonatomic, strong) NSTextField *percentLabel;
@property (nonatomic, strong) NSTextField *timeLabel;

@property (nonatomic, strong) NSTextField *hourLabel;
@property (nonatomic, strong) NSTextField *minuteLabel;
@property (nonatomic, strong) NSTextField *secondLabel;

@end

@implementation TimeFliesView

static NSDateFormatter *_formatter = nil;

- (instancetype)initWithFrame:(NSRect)frame isPreview:(BOOL)isPreview {
    self = [super initWithFrame:frame isPreview:isPreview];
    
    [self setAnimationTimeInterval: 1 / 20.0];
    
    [self prepareForDisplay];
    
    return self;
}

- (void)drawRect:(NSRect)rect {
    [super drawRect:rect];
    
    [[NSColor blackColor] setFill];
    NSRectFill(rect);
}

- (void)animateOneFrame {
    [self update];
}

- (CGFloat)fontSizeWith:(CGFloat)factor {
    CGFloat ref = MIN(self.bounds.size.width, self.bounds.size.height);
    CGFloat result = ref * factor;
    return result;
}

- (void)prepareForDisplay {
    _customLabel = [NSTextField labelWithString:@"Vergil @ Flexport"];
    
    _percentNumber = [NSTextField labelWithString:@"Hello World"];
    _percentNumber.textColor = NSColor.whiteColor;
    
    _percentLabel = [NSTextField labelWithString:@"of this year has passed."];
    _percentLabel.textColor = NSColor.whiteColor;
    
    [self addSubview:_percentNumber];
    [self addSubview:_percentLabel];
    
    if (self.bounds.size.width > self.bounds.size.height) {
        [self prepareForHorizontal];
    } else {
        [self prepareForVertical];
    }
    
    [self update];
}

- (void)prepareForVertical {
    _hourLabel = [NSTextField labelWithString:@"00"];
    _hourLabel.textColor = NSColor.whiteColor;
    
    _minuteLabel = [NSTextField labelWithString:@"00"];
    _minuteLabel.textColor = NSColor.whiteColor;
    
    _secondLabel = [NSTextField labelWithString:@"00"];
    _secondLabel.textColor = NSColor.whiteColor;
    
    [self addSubview:_hourLabel];
    [self addSubview:_minuteLabel];
    [self addSubview:_secondLabel];
}

- (void)prepareForHorizontal {
    _timeLabel = [NSTextField labelWithString:@"00:00:00"];
    _timeLabel.textColor = NSColor.whiteColor;
    
    _formatter = [[NSDateFormatter alloc] init];
    [_formatter setDateFormat:@"HH:mm:ss"];
    
    [self addSubview:_timeLabel];
}

- (void)update {
    if (self.bounds.size.width > self.bounds.size.height) {
        [self updateHorizontal];
    } else {
        [self updateVertical];
    }
}

- (void)updateVertical {
    CGFloat x = 0;
    CGFloat y = 0;
    
    NSDate *now = [NSDate date];
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *components = [calendar components:NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond fromDate:now];
    
    _hourLabel.stringValue = [NSString stringWithFormat:@"%02ld", (long)[components hour]];
    _hourLabel.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.3] weight:NSFontWeightUltraLight];
    [_hourLabel sizeToFit];
    x = self.bounds.size.width / 2 - _hourLabel.frame.size.width / 2;
    y = self.bounds.size.height - self.bounds.size.height * .36;
    _hourLabel.frame = CGRectMake(x, y, _hourLabel.frame.size.width, _hourLabel.frame.size.height);
    
    _minuteLabel.stringValue = [NSString stringWithFormat:@"%02ld", (long)[components minute]];
    _minuteLabel.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.3] weight:NSFontWeightUltraLight];
    [_minuteLabel sizeToFit];
    x = self.bounds.size.width / 2 - _minuteLabel.frame.size.width / 2;
    y = _hourLabel.frame.origin.y - _minuteLabel.frame.size.height - self.bounds.size.height * .01;
    _minuteLabel.frame = CGRectMake(x, y, _minuteLabel.frame.size.width, _minuteLabel.frame.size.height);
    
    _secondLabel.stringValue = [NSString stringWithFormat:@"%02ld", (long)[components second]];
    _secondLabel.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.3] weight:NSFontWeightUltraLight];
    [_secondLabel sizeToFit];
    x = self.bounds.size.width / 2 - _secondLabel.frame.size.width / 2;
    y = _minuteLabel.frame.origin.y - _secondLabel.frame.size.height - self.bounds.size.height * .01;
    _secondLabel.frame = CGRectMake(x, y, _secondLabel.frame.size.width, _secondLabel.frame.size.height);
    
    _percentLabel.font = [NSFont systemFontOfSize:[self fontSizeWith:.01] weight:NSFontWeightSemibold];
    [_percentLabel sizeToFit];
    x = self.bounds.size.width / 2 - _percentLabel.frame.size.width / 2;
    y = self.bounds.size.height * .05;
    _percentLabel.frame = CGRectMake(x, y, _percentLabel.frame.size.width, _percentLabel.frame.size.height);
    
    _percentNumber.stringValue = [NSString stringWithFormat:@"%.06f %%", [self remainPercentOfYear]];
    _percentNumber.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.05] weight:NSFontWeightUltraLight];
    [_percentNumber sizeToFit];
    x = self.bounds.size.width / 2 - _percentNumber.frame.size.width / 2;
    y = _percentLabel.frame.origin.y + _percentLabel.frame.size.height + self.bounds.size.height * .01;
    _percentNumber.frame = CGRectMake(x, y, _percentNumber.frame.size.width, _percentNumber.frame.size.height);
}

- (void)updateHorizontal {
    CGFloat x = 0;
    CGFloat y = 0;
    
    _percentNumber.stringValue = [NSString stringWithFormat:@"%.6f %%", [self remainPercentOfYear]];
    _percentNumber.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.2] weight:NSFontWeightUltraLight];
    [_percentNumber sizeToFit];
    x = self.bounds.size.width / 2 - _percentNumber.frame.size.width / 2;
    y = self.bounds.size.height / 2 - _percentNumber.frame.size.height / 2;
    _percentNumber.frame = CGRectMake(x, y, _percentNumber.frame.size.width, _percentNumber.frame.size.height);
    
    _percentLabel.font = [NSFont systemFontOfSize:[self fontSizeWith:.02] weight:NSFontWeightSemibold];
    [_percentLabel sizeToFit];
    x = self.bounds.size.width / 2 - _percentLabel.frame.size.width / 2;
    y = _percentNumber.frame.origin.y - _percentLabel.frame.size.height;
    _percentLabel.frame = CGRectMake(x, y, _percentLabel.frame.size.width, _percentLabel.frame.size.height);
    
    _timeLabel.stringValue = [_formatter stringFromDate:[NSDate date]];
    _timeLabel.font = [NSFont monospacedDigitSystemFontOfSize:[self fontSizeWith:.06] weight:NSFontWeightUltraLight];
    [_timeLabel sizeToFit];
    x = self.bounds.size.width / 2 - _timeLabel.frame.size.width / 2;
    y = self.bounds.size.height * 0.1;
    _timeLabel.frame = CGRectMake(x, y, _timeLabel.frame.size.width, _timeLabel.frame.size.height);
}

- (double)remainPercentOfYear {
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDate *now = [NSDate date];
    NSDateComponents *yearComponent = [calendar components:NSCalendarUnitYear fromDate:now];
    NSDate *thisYear = [calendar dateFromComponents:yearComponent];
    
    double passed = fabs([thisYear timeIntervalSinceNow]);
    
    NSDateComponents *otherCompoent = [[NSDateComponents alloc] init];
    otherCompoent.calendar = calendar;
    otherCompoent.year = yearComponent.year + 1;
    
    double remainOfThisYear = [[calendar dateFromComponents:otherCompoent] timeIntervalSinceNow];
    
    return passed / (remainOfThisYear + passed) * 100;
}

@end

File: TimeFlies/TimeFliesView.swift
//
//  TimeFliesView.swift
//  TimeFlies
//
//  Created by Vergil Choi on 2018/4/9.
//  Copyright © 2018 Vergil Choi. All rights reserved.
//

import AppKit
import ScreenSaver

class TimeFliesView: ScreenSaverView {
    
    private let percentNumber: NSTextField = {
        let view = NSTextField(labelWithString: "Hello World")
        view.textColor = .white
        return view
    }()
    
    private let percentLabel: NSTextField = {
        let view = NSTextField(labelWithString: "of this year has passed.")
        view.textColor = .white
        return view
    }()

    
    convenience init() {
        self.init(frame: NSRect.zero, isPreview: false)
    }
    
    override init!(frame: NSRect, isPreview: Bool) {
        super.init(frame: frame, isPreview: isPreview)
        
        prepareForDisplay()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        
        prepareForDisplay()
    }
    
//    private lazy var configuration = {
//        return ConfigurationController()
//    }()
//
//    override var hasConfigureSheet: Bool {
//        return true
//    }
//
//    override var configureSheet: NSWindow? {
//        return configuration.window
//    }
    
    override func animateOneFrame() {
        update()
    }
    
    private func prepareForDisplay() {
        update()
        addSubview(percentNumber)
        addSubview(percentLabel)
    }
    
    private func update() {
        percentNumber.stringValue = String(format: "%.6f %%", remainPercentOfYear())
        
        percentNumber.font = NSFont.monospacedDigitSystemFont(ofSize: bounds.height * 0.2, weight: NSFont.Weight.ultraLight)
        percentNumber.sizeToFit()
        percentNumber.frame.origin.x = bounds.width / 2 - percentNumber.frame.width / 2
        percentNumber.frame.origin.y = bounds.height / 2 - percentNumber.frame.height / 2
        
        percentLabel.font = NSFont.systemFont(ofSize: bounds.height * 0.02, weight: NSFont.Weight.semibold)
        percentLabel.sizeToFit()
        percentLabel.frame.origin.x = bounds.width / 2 - percentLabel.frame.width / 2
        percentLabel.frame.origin.y = percentNumber.frame.origin.y - percentLabel.frame.height
        
    }

    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)

        // Drawing code here.
        NSColor.black.setFill()
        dirtyRect.fill()
    }
    
    private func remainPercentOfYear() -> Double {
        
        let calendar = Calendar.current
        let now = Date()
        let yearComponent = calendar.dateComponents([.year], from: now)
        let thisYear = calendar.date(from: yearComponent)!
        
        let passed = abs(thisYear.timeIntervalSinceNow)
        
        let longComponent = DateComponents(calendar: calendar, timeZone: nil, era: nil, year: yearComponent.year! + 1, month: nil, day: nil, hour: nil, minute: nil, second: nil, nanosecond: nil, weekday: nil, weekdayOrdinal: nil, quarter: nil, weekOfMonth: nil, weekOfYear: nil, yearForWeekOfYear: nil)
        let longOfThisYear = calendar.date(from: longComponent)!.timeIntervalSinceNow + passed
        
        return passed / longOfThisYear * 100
    }
    
    
}
